python/cudf/cudf/core/column/decimal.py

```diff
@@ -370,170 +376,20 @@ def as_numerical_column(self, dtype: np.dtype) -> NumericalColumn:
-        # 0 for non-negative values, -1 for negative values:
-        data_buf_128[1::4] = np.piecewise(
-            data_buf_32, [data_buf_32 < 0], [-1, 0]
-        )
-        data_buf_128[2::4] = np.piecewise(
-            data_buf_32, [data_buf_32 < 0], [-1, 0]
-        )
-        data_buf_128[3::4] = np.piecewise(
-            data_buf_32, [data_buf_32 < 0], [-1, 0]
-        )
-        data_buf = pa.py_buffer(data_buf_128)
-        mask_buf = (
-            self.base_mask
-            if self.base_mask is None
-            else pa.py_buffer(self.base_mask.memoryview())
-        )
-        return pa.Array.from_buffers(
-            type=self.dtype.to_arrow(),  # type: ignore[union-attr]
-            offset=self.offset,
-            length=self.size,
-            # PyArrow stubs are too strict - from_buffers should accept None for missing buffers
-            buffers=[mask_buf, data_buf],  # type: ignore[list-item]
-        )
-
-    def _with_type_metadata(self: Self, dtype: DtypeObj) -> Self:
-        if isinstance(dtype, Decimal32Dtype):
-            self.dtype.precision = dtype.precision  # type: ignore[union-attr]
-        if cudf.get_option("mode.pandas_compatible"):
-            self._dtype = get_dtype_of_same_type(dtype, self.dtype)
-        return self
-
-
-class Decimal128Column(DecimalBaseColumn):
-    _VALID_PLC_TYPES = {plc.TypeId.DECIMAL128}
-
-    def __init__(
-        self,
-        plc_column: plc.Column,
-        dtype: Decimal128Dtype,
-        exposed: bool,
-    ) -> None:
-        if (
-            not cudf.get_option("mode.pandas_compatible")
-            and not isinstance(dtype, Decimal128Dtype)
-        ) or (
-            cudf.get_option("mode.pandas_compatible")
-            and not is_decimal128_dtype(dtype)
-        ):
-            raise ValueError(f"{dtype=} must be a Decimal128Dtype instance")
-        super().__init__(
-            plc_column=plc_column,
-            dtype=dtype,
-            exposed=exposed,
-        )
-
-    @classmethod
-    def from_arrow(cls, data: pa.Array | pa.ChunkedArray) -> Self:
-        result = cast(Decimal128Dtype, super().from_arrow(data))
-        result.dtype.precision = data.type.precision
-        return result
-
-    def to_arrow(self) -> pa.Array:
-        dtype: Decimal128Dtype
-        if isinstance(self.dtype, pd.ArrowDtype):
-            dtype = pyarrow_dtype_to_cudf_dtype(self.dtype)  # type: ignore[assignment]
-        else:
-            dtype = self.dtype  # type: ignore[assignment]
-
-        return super().to_arrow().cast(dtype.to_arrow())
-
-    def _with_type_metadata(self: Self, dtype: DtypeObj) -> Self:
-        if isinstance(dtype, Decimal128Dtype):
-            self.dtype.precision = dtype.precision  # type: ignore[union-attr]
-        if cudf.get_option("mode.pandas_compatible"):
-            self._dtype = get_dtype_of_same_type(dtype, self.dtype)
-        return self
+    _decimal_cls = Decimal32Dtype
+    _decimal_check = is_decimal32_dtype
```

Comment:
NIt: Typically I avoid these `is_*_dtype` calls internally because they allow more than just an instances of a `DecimalDtype` (they accept decimal columns, types, etc). Could we change these to `lambda dtype: isinstance(dtype, cls._decimal_cls)` or just inline that call?